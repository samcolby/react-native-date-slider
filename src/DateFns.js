import moment from "moment";

function initLocale(name, config) {
  try {
    moment.locale(name, config);
  } catch (error) {
    throw new Error(
      "Locale prop is not in the correct format. \n Locale has to be in form of object, with keys of name and config. " +
        error.message
    );
  }
}

function initMoment(date, locale) {
  if (locale && locale.name) {
    return moment(date).startOf("day").locale(locale.name);
  } else {
    return moment(date).startOf("day");
  }
}

export default { initLocale, initMoment };
