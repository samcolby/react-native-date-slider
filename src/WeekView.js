import React, { PureComponent } from "react";
import { View } from "react-native";

import CalendarHeader from "./CalendarHeader";
import CalendarDay from "./CalendarDay";

const DAYS_TO_GENERATE_ARRAY = [1, 2, 3, 4, 5, 6];

class WeekView extends PureComponent {
  static propTypes = {
    startingDate: React.PropTypes.object.isRequired,
    arrSelected: React.PropTypes.array.isRequired,
    onDateSelected: React.PropTypes.func.isRequired,
    viewWidth: React.PropTypes.number.isRequired
  };

  constructor(props) {
    super(props);

    this.state = {
      arrDays: this.generateArrDays(this.props.startingDate)
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.startingDate !== this.state.arrDays[0]) {
      this.setState({
        arrDays: this.generateArrDays(nextProps.startingDate)
      });
    }
  }

  /**
   * Generates an array of the days for this week.
   * This uses the passed in startingDate as the first day of the array
   * so as to keep the CalendarDays working as a PureComponent
   *
   * @param {moment} startingDate
   *  The starting date for this week
   * @returns {Array<moment>}
   *  An array of moment instances representing the week
   *
   * @memberof WeekView
   */
  generateArrDays(startingDate) {
    return [
      startingDate,
      ...DAYS_TO_GENERATE_ARRAY.map(i => startingDate.clone().add(i, "days"))
    ];
  }

  render() {
    const { arrDays } = this.state;

    const arrCalendarDays = arrDays.map((date, i) => {
      return (
        <CalendarDay
          key={date.toISOString()}
          date={date}
          onDateSelected={this.props.onDateSelected}
          selected={this.props.arrSelected[i]}
        />
      );
    });

    return (
      <View style={{ flexGrow: 1, height: 100 }}>
        <CalendarHeader startDate={this.props.startingDate} />

        <View
          style={{
            flex: 1,
            flexDirection: "row",
            justifyContent: "space-between",
            paddingHorizontal: 12,
            width: this.props.viewWidth
          }}
        >
          {arrCalendarDays}
        </View>
      </View>
    );
  }
}

export default WeekView;
