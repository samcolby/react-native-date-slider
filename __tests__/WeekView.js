import React from "react";

import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import _ from "lodash";
import moment from "moment";

import DateFns from "../src/DateFns";
import WeekView from "../src/WeekView";

describe("WeekView Component", () => {
  it("should shallow render without issues", () => {
    const startingDate = DateFns.getStartOfWeek(moment.utc("2017-06-18"));
    const arrSelected = [false, true, false, false, false, false, false];
    const onDateSelected = jest.fn();
    const component = shallow(
      <WeekView
        startingDate={startingDate}
        arrSelected={arrSelected}
        onDateSelected={onDateSelected}
        viewWidth={750}
      />
    );

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should render 7 CalendarDays", () => {
    const startingDate = DateFns.getStartOfWeek(moment.utc("2017-06-18"));
    const arrSelected = [false, true, false, false, false, false, false];
    const onDateSelected = jest.fn();
    const component = shallow(
      <WeekView
        startingDate={startingDate}
        arrSelected={arrSelected}
        onDateSelected={onDateSelected}
        viewWidth={750}
      />
    );

    expect(component.find("CalendarDay").length).toBe(7);
  });

  it("should render the correct dates", () => {
    const startingDate = DateFns.getStartOfWeek(moment.utc("2017-06-18"));
    const arrSelected = [false, false, false, false, false, false, false];
    const onDateSelected = jest.fn();
    const component = shallow(
      <WeekView
        startingDate={startingDate}
        arrSelected={arrSelected}
        onDateSelected={onDateSelected}
        viewWidth={750}
      />
    );

    expect(component.find("CalendarDay").length).toBe(7);
    component.find("CalendarDay").forEach((child, j) => {
      const date = startingDate.clone().add(j, "days");
      expect(child.props().date.toISOString()).toBe(date.toISOString());
    });
  });

  _.range(7).forEach(i => {
    it(`should set the correct selectedDate for day ${i + 1}`, () => {
      const startingDate = DateFns.getStartOfWeek(moment.utc("2017-06-18"));
      const arrSelected = [false, false, false, false, false, false, false];
      arrSelected[i] = true;
      const onDateSelected = jest.fn();
      const component = shallow(
        <WeekView
          startingDate={startingDate}
          arrSelected={arrSelected}
          onDateSelected={onDateSelected}
          viewWidth={750}
        />
      );

      expect(component.find("CalendarDay").length).toBe(7);
      component.find("CalendarDay").forEach((child, j) => {
        if (i == j) {
          expect(child.props().selected).toBe(true);
        } else {
          expect(child.props().selected).toBe(false);
        }
      });
    });
  });

  it("should not rerender if the props are the same", () => {
    const startingDate = DateFns.getStartOfWeek(moment.utc("2017-06-18"));
    const arrSelected = [false, false, false, false, false, false, false];
    const onDateSelected = jest.fn();
    const component = shallow(
      <WeekView
        startingDate={startingDate}
        arrSelected={arrSelected}
        onDateSelected={onDateSelected}
        viewWidth={750}
      />
    );

    const firstRender = toJson(component);

    component.setProps({
      startingDate,
      arrSelected,
      onDateSelected,
      viewWidth: 750
    });
    expect(toJson(component)).toEqual(firstRender);
  });

  it("should update if the selected date changes", () => {
    const startingDate = DateFns.getStartOfWeek(moment.utc("2017-06-18"));
    const arrSelected = [true, false, false, false, false, false, false];
    const onDateSelected = jest.fn();
    const component = shallow(
      <WeekView
        startingDate={startingDate}
        arrSelected={arrSelected}
        onDateSelected={onDateSelected}
        viewWidth={750}
      />
    );

    const firstRender = toJson(component);
    arrSelected[0] = false;
    arrSelected[2] = true;

    component.setProps({
      startingDate: startingDate,
      arrSelected: [...arrSelected],
      onDateSelected: onDateSelected,
      viewWidth: 750
    });

    const secondRender = toJson(component);

    expect(secondRender).toMatchSnapshot();
    expect(secondRender).not.toEqual(firstRender);

    component.find("CalendarDay").forEach((child, j) => {
      if (j === 2) {
        expect(child.props().selected).toBe(true);
      } else {
        expect(child.props().selected).toBe(false);
      }
    });
  });

  it("should rerender if the starting date changes", () => {
    const startingDate = DateFns.getStartOfWeek(moment.utc("2017-06-18"));
    const arrSelected = [true, false, false, false, false, false, false];
    const onDateSelected = jest.fn();
    const component = shallow(
      <WeekView
        startingDate={startingDate}
        arrSelected={arrSelected}
        onDateSelected={onDateSelected}
        viewWidth={750}
      />
    );

    const firstRender = toJson(component);

    const newStartingDate = DateFns.getStartOfWeek(moment.utc("2015-06-18"));
    component.setProps({
      startingDate: newStartingDate,
      arrSelected: arrSelected,
      onDateSelected: onDateSelected,
      viewWidth: 750
    });

    const secondRender = toJson(component);

    expect(secondRender).toMatchSnapshot();
    expect(secondRender).not.toEqual(firstRender);

    component.find("CalendarDay").forEach((child, j) => {
      const date = newStartingDate.clone().add(j, "days");
      expect(child.props().date.toISOString()).toBe(date.toISOString());
    });
  });
});
