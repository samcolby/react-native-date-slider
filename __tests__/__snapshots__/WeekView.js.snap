// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeekView Component should rerender if the starting date changes 1`] = `
<View
  style={
    Object {
      "flexGrow": 1,
      "height": 100,
    }
  }
>
  <CalendarHeader
    calendarHeaderFormat="MMMM YYYY"
    startingDate={"2015-06-15T00:00:00.000Z"}
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 12,
        "width": 750,
      }
    }
  >
    <CalendarDay
      date={"2015-06-15T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={true}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2015-06-16T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2015-06-17T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2015-06-18T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2015-06-19T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2015-06-20T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2015-06-21T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
  </View>
</View>
`;

exports[`WeekView Component should shallow render without issues 1`] = `
<View
  style={
    Object {
      "flexGrow": 1,
      "height": 100,
    }
  }
>
  <CalendarHeader
    calendarHeaderFormat="MMMM YYYY"
    startingDate={"2017-06-12T00:00:00.000Z"}
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 12,
        "width": 750,
      }
    }
  >
    <CalendarDay
      date={"2017-06-12T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-13T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={true}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-14T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-15T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-16T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-17T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-18T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
  </View>
</View>
`;

exports[`WeekView Component should update if the selected date changes 1`] = `
<View
  style={
    Object {
      "flexGrow": 1,
      "height": 100,
    }
  }
>
  <CalendarHeader
    calendarHeaderFormat="MMMM YYYY"
    startingDate={"2017-06-12T00:00:00.000Z"}
  />
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingHorizontal": 12,
        "width": 750,
      }
    }
  >
    <CalendarDay
      date={"2017-06-12T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-13T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-14T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={true}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-15T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-16T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-17T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
    <CalendarDay
      date={"2017-06-18T00:00:00.000Z"}
      onDateSelected={[Function]}
      selected={false}
      showDayName={true}
      showDayNumber={true}
      styleWeekend={true}
    />
  </View>
</View>
`;
